odoo/addons/base/models/ir_access.py

```diff
@@ -0,0 +1,424 @@
+from collections.abc import Iterator
+import logging
+
+from odoo import models, fields, api
+from odoo.exceptions import AccessError, ValidationError
+from odoo.osv import expression
+from odoo.tools import ormcache
+from odoo.tools.safe_eval import safe_eval, time
+from odoo.tools.translate import _lt, _
+
+_logger = logging.getLogger(__name__)
+
+OPERATIONS = {
+    'read': _lt("read"),
+    'write': _lt("write"),
+    'create': _lt("create"),
+    'unlink': _lt("unlink"),
+}
+
+MODEL_ACCESS_ERROR = {
+    'read': _lt("You are not allowed to access '%(document_kind)s' (%(document_model)s) records."),
+    'write': _lt("You are not allowed to modify '%(document_kind)s' (%(document_model)s) records."),
+    'create': _lt("You are not allowed to create '%(document_kind)s' (%(document_model)s) records."),
+    'unlink': _lt("You are not allowed to delete '%(document_kind)s' (%(document_model)s) records."),
+}
+
+OPERATION_SELECTION = [
+    ('r', 'Read'),
+    ('w', 'Write'),
+    ('rw', 'Read, Write'),
+    ('c', 'Create'),
+    ('rc', 'Read, Create'),
+    ('wc', 'Write, Create'),
+    ('rwc', 'Read, Write, Create'),
+    ('d', 'Delete'),
+    ('rd', 'Read, Delete'),
+    ('wd', 'Write, Delete'),
+    ('rwd', 'Read, Write, Delete'),
+    ('cd', 'Create, Delete'),
+    ('rcd', 'Read, Create, Delete'),
+    ('wcd', 'Write, Create, Delete'),
+    ('rwcd', 'Read, Write, Create, Delete'),
+]
+
+
+class IrAccess(models.Model):
+    """ Access control records with domains. """
+    _name = 'ir.access'
+    _description = "Access"
+    _order = 'model_id, group_id, id'
+    _allow_sudo_commands = False
+
+    name = fields.Char()
+    active = fields.Boolean(default=True, help="Only active accesses are taken into account when checking access rights.")
+    model_id = fields.Many2one('ir.model', string="Model", required=True, ondelete='cascade', index=True)
+    group_id = fields.Many2one('res.groups', string="Group", ondelete='restrict', index=True)
+    # TODO operation: required maybe? default='rwcd'?
+    operation = fields.Selection(OPERATION_SELECTION, help="Which operation(s) this access applies to, a subset of 'rwcd'.")
```

Comment:
What's the purpose of this?
* You need to remember the order of `rwcd`, which is different from the more common `CRUD`
* You can't use it (in a friendly manner) in domains (i.e. find all the `ir.access` that give a read access)
* You store the information twice